### PR TITLE
Moon monorepo

### PR DESCRIPTION
**Pull request by tbezman**
## Frontend Monorepo 🌕

[Loom Walkthrough](https://www.loom.com/share/0d69da1ef6dd433f9882551f1ddee61a)

[Moonrepo documentation here](https://moonrepo.dev/)

**If you want to test this branch make sure you rm -rf your node modules first**

### Directory Structure
This is the structure recommended by Moon and I think it makes sense for us

- Root (gitignore, nvmrc, readme, package json)
  - apps (all of our user facing apps eg. web & mobile)
  - packages (code shared between our apps eg. frames design system)

### Changes to `web`
We no longer have any `scripts` in our `package.json`. All commands are handled via `moon`. All of the old scripts have equivalent `moon run web:{old-script-here}` tasks.

### Scaffolding the mobile app
I scaffolded the `mobile` app using [`npx create-expo-app`](https://docs.expo.dev/get-started/create-a-new-app/) and made some changes to support monorepos supported by [this guide](https://docs.expo.dev/guides/monorepos/)

I was able to test out that importing and path aliasing worked by creating a temporary typescript file in the frames-dls package and using it in mobile & web. Things worked well.

### Github Actions
**I have not migrated e2e and screenshots** since those are disabled. Should be trivial once they come around.

I renamed our actions a little bit just to make it clear that these existing actions are for the web app.

All of our actions have gotten a lot smaller.
1. Moon handles all of the task dependencies for us
2. We were doing way more than we needed to to persist our yarn cache.

### Gitignore
The `.gitignore` needed some love.
1. The root ignore file ignores things that should be ignored across the whole repo
2. Each app / package has its own specific ignore file

### Typescript Configs
Miles has an insane grip on setting these up in a monorepo. He wrote an excellent guide on managing this that I followed [here](https://moonrepo.dev/docs/guides/javascript/typescript-project-refs)

### Mocking fetch
There was some weirdness around upgrading to Node 19 and mocking `fetch` in our tests. I stumbled through it but using `node-fetch@2` w/ a jest setup file did the trick

### CODE REVIEW

**Comment by Robinnnnn** [comment]
> upgrade to node 18

let's just go full 19? https://nodejs.org/en/download/releases/

File: .moon/toolchain.yml [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+  # Name of `tsconfig.json` file in each project root.
+  # projectConfigFileName: 'tsconfig.json'
+
+  # Name of `tsconfig.json` file in the workspace root.
+  # rootConfigFileName: 'tsconfig.json'
+
+  # Name of the config file in the workspace root that defines shared compiler
+  # options for all project reference based config files.
+  # rootOptionsConfigFileName: 'tsconfig.options.json'
```
Comment (Robinnnnn) [inline .moon/toolchain.yml:60]: any reason these are commented out?

File: .nvmrc [hunk 2]
```diff
@@ -1 +1 @@
-16.12.0
\ No newline at end of file
+19.5.0
```
Comment (Robinnnnn) [inline .nvmrc:1]: 😍 thank you!

File: README.md [hunk 3]
```diff
@@ -9,36 +9,30 @@ The premier gallery experience for your NFTs.
...
 ```bash
-yarn relay
+cp apps/web/.env.sample apps/web/.env
 ```
 
 Start app:
 
 ```bash
-$ yarn dev
+yarn moon run web:dev
```
Comment (Robinnnnn) [inline README.md:35]: here the script is `yarn moon`, but in the `other commands` section below it's just `moon`.

probably worth adding in this readme a recommendation to download moon, and that the docs will assume you have it enabled (and that `yarn` can be used as a backup)

File: apps/mobile/.gitignore [hunk 4]
```diff
@@ -0,0 +1,10 @@
+.expo/
+dist/
+npm-debug.*
+*.jks
+*.p8
+*.p12
+*.key
+*.mobileprovision
```
Comment (Robinnnnn) [inline apps/mobile/.gitignore:8]: sexy

File: apps/mobile/package.json [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  "dependencies": {
+    "expo": "~47.0.12",
+    "expo-status-bar": "~1.4.2",
+    "react": "18.1.0",
+    "react-native": "0.70.5"
+  },
+  "devDependencies": {
+    "@babel/core": "^7.12.9",
+    "@types/react-native": "^0.71.1",
+    "babel-plugin-tsconfig-paths-module-resolver": "^1.0.4"
```
Comment (Robinnnnn) [inline apps/mobile/package.json:14]: should probably lock these down in mobile?

File: apps/mobile/moon.yml [hunk 6]
```diff
@@ -0,0 +1,11 @@
+language: 'javascript'
+
+tasks:
+  typecheck:
+    command: tsc --build
+
+  android:
+    command: expo start --ios
```
Comment (Robinnnnn) [inline apps/mobile/moon.yml:8]: android start flag is `--ios`?

File: apps/web/moon.yml [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    - yarn moon run web:graphql-codegen-watch
+
+  lint:
+    command:
+    - eslint
+    - --ext
+    - .js,.ts,.tsx,.cjs
+    - '@group(lint)'
+    inputs:
+      - '@group(lint)'
```
Comment (Robinnnnn) [inline apps/web/moon.yml:44]: normal to have the same command and inputs?

File: apps/web/src/contexts/globalLayout/GlobalLayoutContext.tsx [hunk 8]
```diff
@@ -16,7 +16,6 @@ import { useFragment, useLazyLoadQuery } from 'react-relay';
 import { fetchQuery, graphql } from 'relay-runtime';
 import styled from 'styled-components';
 
-import NavLink from '~/components/core/NavLink/NavLink';
```
Comment (Robinnnnn) [inline apps/web/src/contexts/globalLayout/GlobalLayoutContext.tsx:19]: massive storage savings

File: apps/web/tests/setupFile.js [hunk 9]
```diff
@@ -0,0 +1,2 @@
+import fetch from 'node-fetch';
+global.fetch = fetch;
```
Comment (Robinnnnn) [inline apps/web/tests/setupFile.js:2]: great catch

**Review by Robinnnnn — APPROVED** [review]
overall, i'm really fucking excited for this update. L F GGGGGG 🔥🔥🔥 

thanks for recording the loom as well. lots of gold in here- remote caching sounds gorgeous.

my only slight hesitation would be that we're using a toolchain that's still pre-v1, and we are not yet fully aware of what kind of complexities may arise as we lean deeper into the RN development cycle. at the same time, @milesj is a freaking monster, so...

**Comment by milesj** [comment]
@Robinnnnn It's pretty stable :P v1 will be landing soon!

File: .moon/toolchain.yml (hunk 1 above)
Comment (tbezman) [reply .moon/toolchain.yml:60]: it's what moon scaffolds us with. I feel like it's nice to have these comments. wdyt?

File: README.md (hunk 3 above)
Comment (tbezman) [reply README.md:35]: Updated the readme to talk more about this

File: apps/mobile/package.json (hunk 5 above)
Comment (tbezman) [reply apps/mobile/package.json:14]: I'me fine with patch upgrades happening automatically

File: apps/web/moon.yml (hunk 7 above)
Comment (tbezman) [reply apps/web/moon.yml:44]: I'm pretty sure this is normal? cc @milesj

Comment (milesj) [reply apps/web/moon.yml:44]: Yeah, they just get expanded.

File: .moon/toolchain.yml (hunk 1 above)
Comment (milesj) [reply .moon/toolchain.yml:60]: These are the defaults anyways.

**Pull request merged**